cdc_kafka/kafka.py

```diff
@@ -0,0 +1,120 @@
+import collections
+import datetime
+import json
+import logging
+import socket
+from typing import List, Dict
+
+import avro
+import confluent_kafka.avro
+
+logger = logging.getLogger(__name__)
+
+
+class KafkaClient(object):
+    def __init__(self, bootstrap_servers: str, schema_registry_url: str, kafka_timeout_seconds: int,
+                 extra_kafka_consumer_config: List[str], extra_kafka_producer_config: List[str]):
+        consumer_config = {'bootstrap.servers': bootstrap_servers,
+                           'group.id': f'cdc_to_kafka_progress_check_{socket.getfqdn()}',
+                           'schema.registry.url': schema_registry_url}
+        producer_config = {'bootstrap.servers': bootstrap_servers,
+                           'linger.ms': '20',
+                           'schema.registry.url': schema_registry_url}
+
+        for extra in extra_kafka_consumer_config or []:
+            k, v = extra.split(':')
+            consumer_config[k] = v
+
+        for extra in extra_kafka_producer_config or []:
+            k, v = extra.split(':')
+            producer_config[k] = v
+
+        logger.debug('Kafka consumer configuration: \n%s', json.dumps(consumer_config, indent=4))
+        logger.debug('Kafka producer configuration: \n%s', json.dumps(producer_config, indent=4))
+
+        self.consumer = confluent_kafka.avro.AvroConsumer(consumer_config)
+        self.producer = confluent_kafka.avro.AvroProducer(producer_config)
+
+        self.kafka_timeout_seconds = kafka_timeout_seconds
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, value, traceback):
+        logger.info("Closing Kafka consumer and flushing Kafka producer...")
+        self.consumer.close()
+        self.producer.flush(self.kafka_timeout_seconds)
+        logger.info("Done.")
+
+    delivery_success_logging_counter = 1000  # to get it to log the first success; will be every 1,000th afterwards
+    last_producer_poll_time = datetime.datetime.now()
+
+    @staticmethod
+    def on_delivery_cb(err: confluent_kafka.KafkaError, msg: confluent_kafka.Message) -> None:
+        if err is not None:
+            logger.error(f'Message delivery failed for topic {msg.topic()} with error: {err}')
+        else:
+            KafkaClient.delivery_success_logging_counter += 1
+            if KafkaClient.delivery_success_logging_counter >= 1000:
```

Comment:
I prefer `%` for this sort of thing but it's fine.

Reply:
Good thought, will change.